Introduce enrich options to handle ECS passthrough. (#464)
logstash-plugins/logstash-input-beats: docs/index.asciidoc

[id="plugins-{type}s-{plugin}-ecs_metadata"]

> ==== Event Enrichment and the Elastic Common Schema (ECS)

Reviewer: Elastic standard: sentence case for headings.
Headings are inconsistent throughout, but we're trying to correct inconsistencies when we get a chance.

> When decoding {plugin-uc} events, this plugin _enriches_ each event with metadata to make information about where the event came from available during further processing.

Reviewer: ```suggestion
When decoding {plugin-uc} events, this plugin enriches each event with metadata about the event's source, making this information available during further processing. 
```

> The <<plugins-{type}s-{plugin}-enrich>> option can be used to activate or deactivate individual enrichment categories.

Reviewer: ```suggestion
You can use the <<plugins-{type}s-{plugin}-enrich>> option to activate or deactivate individual enrichment categories.
```

Reviewer: Use active voice to put the focus back on the user.

The location of these enrichment fields depends on whether <<plugins-{type}s-{plugin}-ecs_compatibility,ECS compatibility mode>> is enabled:

> - When ECS compatibility is _enabled_, enrichment fields will be added in an ECS-compatible structure.
> - When ECS compatibility is _disabled_, enrichment fields will be added in a way that is backward-compatible with this plugin, but is known to clash with the Elastic Common Schema.

Reviewer: ```suggestion
- When ECS compatibility is _enabled_, enrichment fields are added in an ECS-compatible structure.
- When ECS compatibility is _disabled_, enrichment fields are added in a way that is backward-compatible with this plugin, but is known to clash with the Elastic Common Schema.
```

.`source_metadata`

> [cols="<l,<l,<e",caption="Enrichment Category:"]

Reviewer: ```suggestion
[cols="<l,<l,<e",caption="Enrichment category:"]
```

|=======================================================================
|ECS `v1`, `v8` |ECS `disabled` |Description

|[@metadata][input][beats][host][name]
|[host]

> |Name or address of the {plugin-singular} host

Reviewer: Yay! Bonus points for using the attribute. :-)

|[@metadata][input][beats][host][ip]
|[@metadata][ip_address]
|IP address of the {plugin-uc} client that connected to this input
|=======================================================================

> .`ssl_peer_metadata``

Author: Seems extra back tick we have
```suggestion
.`ssl_peer_metadata`
```

> [cols="<l,<l,<e",caption="Enrichment Category:"]

Reviewer: ```suggestion
[cols="<l,<l,<e",caption="Enrichment category:"]
```

|=======================================================================
|ECS `v1`, `v8` |ECS `disabled` |Description

|[@metadata][tls_peer][status]
|[@metadata][tls_peer][status]
|Contains "verified" or "unverified" label; available when SSL is enabled.

|[@metadata][input][beats][tls][version_protocol]
|[@metadata][tls_peer][protocol]

> |Contains the TLS version used (e.g. `TLSv1.2`); available when SSL status is "verified"

Reviewer: ```suggestion
|Contains the TLS version used (such as `TLSv1.2`); available when SSL status is "verified"
```

|[@metadata][input][beats][tls][client][subject]
|[@metadata][tls_peer][subject]

> |Contains the identity name of the remote end (e.g. `CN=artifacts-no-kpi.elastic.co`); available when SSL status is "verified"

Reviewer: ```suggestion
|Contains the identity name of the remote end (such as `CN=artifacts-no-kpi.elastic.co`); available when SSL status is "verified"
```

|[@metadata][input][beats][tls][cipher]
|[@metadata][tls_peer][cipher_suite]

> |Contains the name of cipher suite used (e.g. `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`); available when SSL status is "verified"

Reviewer: ```suggestion
|Contains the name of cipher suite used (such as `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`); available when SSL status is "verified"
```

|=======================================================================

.`codec_metadata`

> [cols="<l,<l,<e",caption="Enrichment Category:"]

Reviewer: ```suggestion
[cols="<l,<l,<e",caption="Enrichment category:"]
```

|=======================================================================
|ECS `v1`, `v8` |ECS `disabled` |Description

|[tag]
|[tag]

> |Contains `beats_input_codec_XXX_applied` where `XXX` is the name of the codec

Reviewer: Remember that anything you hardcode will show up in the Elastic Agent doc as well.  As this is just an example, consider renaming the example to something generic that would work for both Beats and Agent inputs.

|[event][original]
e|N/A

> |When ECS is enabled, this plugin's *default codec* will ensure that the event's `[event][original]` field is populated, using the bytes as-processed by the codec if the field does not already exist on the event being processed.

Reviewer: ```suggestion
|When ECS is enabled, this plugin's *default codec* ensures that the event's `[event][original]` field is populated. If the field does not already exist on the event being processed, the field uses the bytes as-processed by the codec.
```

Reviewer: This sentence was confusing. I made a a guess, but I'm not sure I got it right. 
When sentences are long and convoluted, they become harder to mentally parse and understand. Breaking the content into two or more sentences is a reliable remedy.

Author: I think this simplification makes more sense.

```suggestion
|When ECS is enabled, even if `[event][original]` field does not already exist on the event being processed, this plugin's *default codec* ensures that the field is populated using the bytes as-processed.
```

Author: Thanks @karenzone for the review. I have reviewed and applied your feedbacks/suggestions, simplified this sentence for your final review.

| <<plugins-{type}s-{plugin}-ecs_compatibility>> | <<string,string>>|No
| <<plugins-{type}s-{plugin}-executor_threads>> |<<number,number>>|No
| <<plugins-{type}s-{plugin}-host>> |<<string,string>>|No

> | <<plugins-{type}s-{plugin}-include_codec_tag>> |<<boolean,boolean>>|__Deprecated__

Reviewer: More bonus points for correctly notating deprecations!  🎉

| <<plugins-{type}s-{plugin}-port>> |<<number,number>>|Yes
| <<plugins-{type}s-{plugin}-ssl>> |<<boolean,boolean>>|No
| <<plugins-{type}s-{plugin}-ssl_certificate>> |a valid filesystem path|No
| <<plugins-{type}s-{plugin}-ssl_certificate_authorities>> |<<array,array>>|No
| <<plugins-{type}s-{plugin}-ssl_handshake_timeout>> |<<number,number>>|No
| <<plugins-{type}s-{plugin}-ssl_key>> |a valid filesystem path|No
| <<plugins-{type}s-{plugin}-ssl_key_passphrase>> |<<password,password>>|No
| <<plugins-{type}s-{plugin}-ssl_peer_metadata>> |<<boolean,boolean>>|__Deprecated__
| <<plugins-{type}s-{plugin}-ssl_supported_protocols>> |<<array,array>>|No
| <<plugins-{type}s-{plugin}-ssl_verify_mode>> |<<string,string>>, one of `["none", "peer", "force_peer"]`|No
| <<plugins-{type}s-{plugin}-tls_max_version>> |<<number,number>>|__Deprecated__
| <<plugins-{type}s-{plugin}-tls_min_version>> |<<number,number>>|__Deprecated__

> | <<plugins-{type}s-{plugin}-enrich>> |<<string,string>>|No

Reviewer: Should appear alphabetically in both the list table and description.

Author: TIL: I will do it.

                      disable `ecs_compatibility` for this plugin.
| source_metadata   | Information about the _source_ of the event, such as the IP address
                      of the inbound connection this input received the event from and the
                      name of the Logstash host that processed the event
| ssl_peer_metadata | Detailed information about the _SSL peer_ we received the event from,
                      such as identity information from the SSL client certificate that was
                      presented when establishing a connection to this input
| all               | _alias_ to include _all_ available enrichments (including additional
                      enrichments introduced in future versions of this plugin)
| none              | _alias_ to _exclude_ all available enrichments. Note that, _explicitly_
                      defining <<codec>> with this option will not disable the `ecs_compatibility`,
                      instead it relies on pipeline or codec `ecs_compatibility` configuration.
|=======================================================================

> **Example:**
> This configuration disables _all_ enrichments:

Reviewer: ```suggestion
**Example:**

This configuration disables _all_ enrichments:
```

The minimum TLS version allowed for the encrypted connections.
The value must be one of the following: 1.1 for TLS 1.1, 1.2 for TLS 1.2, 1.3 for TLS 1.3

[id="plugins-{type}s-{plugin}-enrich"]

> ===== `enrich`

Author: We [tried to design the config](https://github.com/logstash-plugins/logstash-input-elastic_agent/issues/3#issuecomment-1343231583) to be "open to extend" in the future. I am okay to rename and give narrowband but one thing if we know the non-metadata enrichment use case, we would have more ideas how to optimally design the flow.

> ===== `enrich`

Reviewer 2: ack! I'm fine using the naming.

* Value type is <<array,array>>

By default, the plugin initializes ECS-disabled plain codec, source and codec metadata such as source host or `[event][original]` will not be added.
`enrich` option helps to manage which metadata to be included. Available options are:

[cols="<,<"]
|=======================================================================
|Option |Description

|`none`                  |Not to include any data
|`default` alias | `codec_metadata` and `codec_metadata` will be included

> |`all` alias | All `codec_metadata`, `codec_metadata` and `ssl_peer_metadata` will be included

Reviewer 2: Do we really need the `default` or `all` aliases? I would argue that we should completely drop them. So that:

1. The documentation should highlight that the default behavior (when the `enrich` option is NOT provided) is to default to `codec_metadata` and `ssl_peer_metadata`.
2. If the user wants to disable enrichment OR override the default behavior, then they can add it.

As such, having a `default` option makes no sense to me because not providing the option is already default. It would just be repetitive and IMHO, useless.

As far as the `all` alias, while I understand that we would be making it easier for users to benefit from all the metadata additions, it would also potentially be a reason why customers change the settings after upgrade if we add a new metadata option. Instead, users should be able to *opt-in* into that option if they have manually changed the `enrich` option.

Reviewer 2: Also, it is fine to drop the "default" alias, e.g., it doesn't make sense to do:

enrich => ["default"]

Reviewer 2: After chatting with @yaauie, and going over the PR the solution for this is to:

1. Make documentation that clearly describes what the default behavior is *when not providing the option*
2. The `default` option, should be an internal thing, not something it makes sense for users to set (e.g. enrich => default). It brings no value to have that as an option users can change. To simplify, we should only make doc changes to accordingly describe the default behavior.
3. I'm fine keeping `all` as a user option.

Author: Thanks for the feedback @roaksoax and it makes sense and I have updated the logic.

> [id="plugins-{type}s-{plugin}-enrich"]
> ===== `enrich`
> * Value type is <<array,array>>
> 
> By default, the plugin initializes ECS-disabled plain codec, source and codec metadata such as source host or `[event][original]` will not be added.
> `enrich` option helps to manage which metadata to be included. Available options are:
> 
> [cols="<,<"]
> |=======================================================================
> |Option |Description
> 
> |`none`                  |Not to include any data
> |`default` alias | `codec_metadata` and `codec_metadata` will be included
> |`all` alias | All `codec_metadata`, `codec_metadata` and `ssl_peer_metadata` will be included
> |`source_metadata` | Source metadata such as host name will be included
> |`codec_metadata` | Codec metadata such as `tag`, `[event][original]` will be included
> |`ssl_peer_metadata` | SSL peer metadata will be included
> |=======================================================================

Reviewer 3: There is some nuance here with how `enrich` interacts with the ECS Compatibility fields table at the top of this doc that call out "availability" as "always" etc. I'll take a stab at re-tabling that info separately, but below we have an in-list table (asciidoc: bare `+` on a newline causes the following block element to be injected to the current list item)

```suggestion
[id="plugins-{type}s-{plugin}-enrich"]
===== `enrich`

  * Value type is <<string,string>>
  ** A <<list,list>> can also be provided
  ** Configures which enrichments are applied to each event
  ** Supported values are:
+
[cols="2l,5"]
|=======================================================================
|Enrichment | Description

| codec_metadata    | Information about how the codec transformed a sequence of bytes into
                      this Event, such as _which_ codec was used. If no <<codec>> is explicitly
                      specified, _excluding_ `codec_metadata` from `enrich` will prevent the
                      default codec itself from adding metadata such as `[event][original]`.
| source_metadata   | Information about the _source_ of the event, such as the IP address
                      of the inbound connection this input received the event from and the
                      name of the Logstash host that processed the event
| ssl_peer_metadata | Detailed information about the _SSL peer_ we received the event from,
                      such as identity information from the SSL client certificate that was
                      presented when establishing a connection to this input
| default           | _alias_ resolving to _at least_ `[codec_metadata, source_metadata]`
| all               | _alias_ to include _all_ available enrichments (including additional
                      enrichments introduced in future versions of this plugin)
| none              | _alias_ to _exclude_ all available enrichments
|=======================================================================
  ** Default value is `default`, which is an alias for for `[codec_metadata, source_metadata]` that may be extended in future versions of this plugin to include additional by-default enrichments.

```

As an example, following configuration enables source and metadata with ECS enabled plain codec:
```
input {
  beats {
    port => 5044

>     enrich => ["default"]

Author: Applied the change and now `enrich` can be provided as a you mentioned string or list of strings.

The minimum TLS version allowed for the encrypted connections.
The value must be one of the following: 1.1 for TLS 1.1, 1.2 for TLS 1.2, 1.3 for TLS 1.3

[id="plugins-{type}s-{plugin}-enrich"]
===== `enrich`

  * Value type is <<string,string>>
  ** A <<list,list>> can also be provided
  ** Configures which enrichments are applied to each event
  ** Supported values are:
+
[cols="2l,5"]
|=======================================================================

> |Enrichment | Description

Author: @karenzone I would appreciate if you take a quick look at the changes. Thanks.

| codec_metadata    | Information about how the codec transformed a sequence of bytes into

>                       this Event, such as _which_ codec was used. If no <<codec>> is explicitly
>                       specified, _excluding_ `codec_metadata` from `enrich` will prevent the
>                       default codec itself from adding metadata such as `[event][original]`.

Reviewer 4: ```suggestion
                      this Event, such as _which_ codec was used. Also, if no <<codec>> is explicitly
                      specified, _excluding_ `codec_metadata` from `enrich` will disable `ecs_compatibility` for this plugin.
```

The minimum TLS version allowed for the encrypted connections.
The value must be one of the following: 1.1 for TLS 1.1, 1.2 for TLS 1.2, 1.3 for TLS 1.3

[id="plugins-{type}s-{plugin}-enrich"]

> ===== `enrich`

Reviewer: Should appear alphabetically in both the list table and description.

                      disable `ecs_compatibility` for this plugin.
| source_metadata   | Information about the _source_ of the event, such as the IP address
                      of the inbound connection this input received the event from and the
                      name of the Logstash host that processed the event
| ssl_peer_metadata | Detailed information about the _SSL peer_ we received the event from,
                      such as identity information from the SSL client certificate that was
                      presented when establishing a connection to this input
| all               | _alias_ to include _all_ available enrichments (including additional
                      enrichments introduced in future versions of this plugin)
| none              | _alias_ to _exclude_ all available enrichments. Note that, _explicitly_
                      defining <<codec>> with this option will not disable the `ecs_compatibility`,
                      instead it relies on pipeline or codec `ecs_compatibility` configuration.
|=======================================================================

>   ** Default value is `[codec_metadata, source_metadata]` that may be extended in future versions of this plugin to include additional enrichments.

Reviewer: Please consider moving this bullet higher in the list (after "Configures which enrichments."  It looks like a mistake floating about on its own.

> As an example, following configuration disables _all_ enrichments:

Reviewer: ```suggestion
**Example:**
This configuration disables _all_ enrichments:
```
